### PR TITLE
Feature/extend interactive context

### PR DESCRIPTION
**Pull request by kfwilson**
Adding an add components method to InteractiveContext that Abie had patched in in ceam experiments

Adding type hints because I got a little confused trying to extend these in public health when I thought plugin config could be more types than it is

### CODE REVIEW

File: src/vivarium/framework/engine.py [hunk 1]
```diff
@@ -57,6 +59,7 @@ def setup(self):
         self.time_step_emitters = {k: self.builder.event.get_emitter(k) for k in self.time_step_events}
         self.end_emitter = self.builder.event.get_emitter('simulation_end')
         self.builder.event.get_emitter('post_setup')(None)
+        self._setup = True
```
Comment (collijk) [inline src/vivarium/framework/engine.py]: Move this to before you emit post setup so components don't try to do anything weird during that lifecycle phase.

File: src/vivarium/framework/engine.py [hunk 2]
```diff
@@ -79,6 +82,11 @@ def finalize(self):
     def report(self):
         return self.values.get_value('metrics')(self.population.population.index)
 
+    def add_components(self, component_list):
```
Comment (collijk) [inline src/vivarium/framework/engine.py]: Replace the call on line 51 with a call to this method.

File: src/vivarium/interface/interactive.py [hunk 3]
```diff
@@ -116,22 +117,24 @@ def replace_component(self, old_component, new_component):
...
+                          plugin_config: Mapping=None) -> InteractiveContext:
     config = build_simulation_configuration()
     config.update(input_config)
     plugin_manager = PluginManager(plugin_config)
 
     return InteractiveContext(config, components, plugin_manager)
 
 
-def setup_simulation(components, input_config=None, plugin_config=None):
+def setup_simulation(components: List, input_config: Union[Mapping, ConfigTree]=None,
```
Comment (collijk) [inline src/vivarium/interface/interactive.py]: The ConfigTree is a Mapping

File: src/vivarium/interface/interactive.py [hunk 4]
```diff
@@ -116,22 +117,24 @@ def replace_component(self, old_component, new_component):
...
+                     plugin_config: Mapping=None) -> InteractiveContext:
     simulation = initialize_simulation(components, input_config, plugin_config)
     simulation.setup()
 
     return simulation
 
 
-def initialize_simulation_from_model_specification(model_specification_file):
+def initialize_simulation_from_model_specification(model_specification_file: str) -> InteractiveContext:
```
Comment (collijk) [inline src/vivarium/interface/interactive.py]: We should probably only have one place that returns an actual Simulation context.  Drop line 144 and instead return `initialize_simulation(components, simulation_config, plugin_config)`

Comment (kfwilson) [reply src/vivarium/interface/interactive.py]: I need the plugin manager to get the parser for components though.

Comment (collijk) [reply src/vivarium/interface/interactive.py]: Gross.

**Pull request merged**